README.md
# ABOUT CherryBlog


## Introduction

CherryBlog implements a very simple blogging website.

- Posts: the main content of the blog
- Pages: the extra static pages like the 'about'
- Tags: each post has one or multiple tags
- Search: word based search in posts and pages
- Caching: in memory caching of data and content
- Drafts: drafts are pages that are not published yet, only visible in development
- URL rerouting: the final URL is not necessarily the path to the file 

At this moment, it has the basic news of the project where each article is a post about the new version that was released.

### Goals

As a developer, I want to share things with the world and run a blog. Finding the right blogging software is difficult if you limit the choices for the final deployment.

- (Semi-)static website that can be hosted anywhere behind an Apache or Nginx server. Or even in a S3 bucket for that matter.
- No database to hold the content _(see note)_.  Data must come from local files only.
- Easy to write content in for instance MarkDown.
- Extensible with themes, features, ...

_Note: No database if the site is very small. If there would be a need for a higher performance on features like searching, it can be included._

#### Inspiration

I started using JBake quite some time ago, but the project does not seem very lively.  I ran two sites with it and I was very pleased with it.   

Furthermore, I have switched from the Java (-alikes) world to Python.  In Python, the only thing I tried was Pelican.  Pelican however was not what I needed for another website (my personal site as freelancer) and I found CherryPy as framework.  I was able to make my personal site in just a matter of hours (the content and template was already available from JBake).  When using CherryPy, I saw a possibility to write something similar like JBake in Python.  

For now, I will not create a static website (like JBake and Pelican do), but I will run a website with data coming from Markdown files, YAML files, ...  In a later phase, I want to build the site to a fully static website, with a backend for the dynamic data like comments on posts.  

## Architecture

### Python 

The project is written in Python 3 and uses the CherryPy framework as backend. Hence the name of the project: CherryBlog.  

Python libraries included:

- cherrypy
- jinja2
- pyyaml
- markdown

### Content
The content of the blog is written in MarkDown. On top of each content file, there's a meta data section containing:

- author
- date written
- image on top of the page
- summary of the content
- tags

The whole content is served from local files, so no need for a database like for instance WordPress, Drupal or Joomla.

### Settings

The settings (or configuration) are done in YAML files.  Multiple environments can be used for development and production sites, each with their own settings.

### Theme 

The default theme is based on two themes found on https://startbootstrap.com/templates/

- [https://startbootstrap.com/templates/blog-post/](https://startbootstrap.com/templates/blog-post/)
- [https://startbootstrap.com/templates/blog-home/](https://startbootstrap.com/templates/blog-home/)

The theme is implemented in Jinja2 templates.

## Open source

### License

The project is open sourced under the MIT license and the sources can be found on Github:   

[https://github.com/vindevoy/cherryblog](https://github.com/vindevoy/cherryblog)

At this moment, I (vindevoy) am the only contributor.  I started the project just end of March 2020.

### Version(s)

The current version is depicted in the version widget.  The code is tagged on Github for each released version.

### Roadmap

A set of issues have been logged in Github.  They are all in backlog for now as this is a non-funded project.  The updates are scheduled:

- new releases: each 2 weeks (will update version to x.Y+1.0)
- bug fixes: 2 days after the release (will update version x.y.Z+1)

Each "sprint", a number of issues will be moved from the backend to the new release.

The major release number will be bumped when a major new feature is introduced.  For instance:

- Implementation of the the mongodb backend
- Static site generation


# DOCUMENTATION


## WARNING

**This project is still rather new and is only in version 1.x. Hence the documentation is very short.  It is provided for people who really want to dig into the project.**

## Installation

For now, the only way to install the software is to clone the Git repository: [https://github.com/vindevoy/cherryblog](https://github.com/vindevoy/cherryblog)

### Requirements

Operating system: 

- Linux or Mac OSX.  
- Windows: Sorry, I do not yet support Windows.

Python:

- Python 3 is required.  

Other software:

- Git (optional): to clone the repository. Alternatively you can download the software as a zip.
- Make (optional): to run the basics commands. You can run without make if you type the commands found in the Makefile.

### Downloading libraries

A Makefile is provided.  Run it using make.

```
make dependencies
```

### Running the server

There are 2 methods to run the server at this moment:

- on the localhost for development
- on a production alike environment

You can add environments copying the files to YOUR_ENVIRONMENT.yml in the environments directory (/src/data/environments) and specifying the environment on the command line:

```
python3 ./src/application/serve.py --env YOUR_ENVIRONMENT
```

#### Running development on localhost

By default, you are running on localhost, allowing the most efficient way to debug.

```
make develop
```

#### Running the site 'in production'

Rename the production.sample file into production.yml in the same directory and adapt the settings as needed. If you want to run on port 80, you must run as root (or via sudo). This is unsafe and it's better to use a proxy like Nginx or Apache.

```
make production
```


# CREDITS


Many thanks to everybody who wrote software that I'm using.  Thanks also to the photographers who share their work.

## Language

CherryBlog is written in Python 3: 

- [https://www.python.org/](https://www.python.org/) 

## Python libraries

CherryBlog uses the following Python libraries:

- cherrypy: [https://cherrypy.org/](https://cherrypy.org/)
- markdown: [https://pypi.org/project/Markdown/](https://pypi.org/project/Markdown/)
- pyyaml: [https://pyyaml.org/](https://pyyaml.org/)
- jinja2: [https://pypi.org/project/Jinja2/](https://pypi.org/project/Jinja2/)

# Templating engine

- Jinja2: [https://jinja.palletsprojects.com/en/2.11.x/](https://jinja.palletsprojects.com/en/2.11.x/)

## Bootstrap template

The default template is based on the startbootstrap.com templates downloaded from:

- [https://startbootstrap.com/templates/blog-home/](https://startbootstrap.com/templates/blog-home/)
- [https://startbootstrap.com/templates/blog-post/](https://startbootstrap.com/templates/blog-post/)

## Images

Images downloaded:

- <a href="https://pixabay.com/users/jingoba-24598/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=212601">Jiří Rotrekl</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=212601">Pixabay</a>
- <a href="https://pixabay.com/photos/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=839864">Free-Photos</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=839864">Pixabay</a>
- <a href="https://pixabay.com/users/ulleo-1834854/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3477927">Ulrike Leone</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3477927">Pixabay</a>
- <a href="https://pixabay.com/users/Couleur-1195798/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=1437707">Diese lizenzfreien Fotos darfst du zwar verwenden</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=1437707">Pixabay</a>
- <a href="https://pixabay.com/users/Hans-2/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=324175">Hans Braxmeier</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=324175">Pixabay</a>
- <a href="https://pixabay.com/users/Hans-2/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=1260641">Hans Braxmeier</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=1260641">Pixabay</a>
- <a href="https://pixabay.com/users/webandi-1460261/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3308735">Andreas Lischka</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3308735">Pixabay</a>
- <a href="https://pixabay.com/users/gayulo-3585927/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3238564">gayulo</a> from <a href="https://pixabay.com/?utm_source=link-attribution&amp;utm_medium=referral&amp;utm_campaign=image&amp;utm_content=3238564">Pixabay</a>

SVG downloaded for logo:

- <a href="https://commons.wikimedia.org/wiki/File:Fruit-cherries.svg?uselang=fr">https://commons.wikimedia.org/wiki/File:Fruit-cherries.svg?uselang=fr</a>

## Inspiration

- JBake: [https://jbake.org/](https://jbake.org/)
- Pelican: [https://blog.getpelican.com/](https://blog.getpelican.com/)

HISTORY.md
# Release v1.6.0


- author: Yves Vindevogel (vindevoy)
- date: 2020-05-17

## New Features

- URL rerouting has been implemented on the level of the posts, pages and tags.  This means that you can change the URL of the publication and through the rerouting, it will map the correct document(s).  This can be used to make URLs more readable.
- Support for draft pages has been included.  If you are writing a new post or page, but you don't want it to be published yet, you can put it as draft in the meta data.  You can decide through the environment.yml if you want to show drafts or not (development vs. production).  

## Enhancements

- A sitemap.xml file can be generated from the Makefile.  This allows Google to crawl the website better.
- The application controller class now uses htmlmin to remove all the comment in the HTML, remove double spaces and so on.  This results in a smaller HTML going over the wire, but it also results in a better memory footprint for caching.

## Other

- A Facebook page has been created for marketing reasons.  Why else ?

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/13](https://github.com/vindevoy/cherryblog/milestone/13)


# Release v1.5.2


- author: Yves Vindevogel (vindevoy)
- date: 2020-05-08

## Hotfix

- There was a runtime-error on the posts and pages if they had to tags. This resulted in a TypeError, and in code the KeyError was captured.  

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/15](https://github.com/vindevoy/cherryblog/milestone/15)


# Release v1.5.1


- author: Yves Vindevogel (vindevoy)
- date: 2020-05-07

## Bug Fixes

- There was a problem with the search results. The font weight (bold) of found text in the search results is now case-insensitive.  

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/14](https://github.com/vindevoy/cherryblog/milestone/14)


# Release v1.5.0


- author: Yves Vindevogel (vindevoy)
- date: 2020-05-01

## Features

- Search widget, with minimal number of characters
- Word(s) based search on all the pages and posts
- Case-insensitive search
- Search results with highlight of the word searched
- Number of occurrences of the searched word(s)
- Author and date of each search result
- Image on top of the page is the image of the first result found
- Uses cache for during search, but does not cache itself
- Exclude pages or posts from the search

## Enhancements

- Default settings for items not in the environment files
- Ability to switch of caching (localhost during development)
- Tags are shown on the bottom of the page or post
- Dates are now written in a nicer format
- Better passing on of data through-out the templates
- Commenting in the templates

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/12](https://github.com/vindevoy/cherryblog/milestone/12)


# Release v1.4.2


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-26

## Enhancements

- Added a simple re-route script to re-route HTTP to HTTPS. Can be used if you don't want to run behind an NGINX or Apache.

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/11](https://github.com/vindevoy/cherryblog/milestone/11)


# Release v1.4.1


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-25

## Fixes

- Colour of the background for code is now aligned with the colour of the boxes
- Blue border is gone on the buttons on the home page (primary buttons)
- Margin problem on the tags for smaller screens
- No data showing in the version widget on smaller screens
- External links go in a new tab or window
- Tags now also display page x of y
- MIT License is linked to a Wikipedia page

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/10](https://github.com/vindevoy/cherryblog/milestone/10)


# Release v1.4.0


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-25

## Enhancements

- Added the code to start with HTTPS instead of HTTP
- CherryBlog.org will run on HTTPS from now on

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/9](https://github.com/vindevoy/cherryblog/milestone/9)


# Release v1.3.0


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-23

## Enhancements

- Page caching added, next to the data caching (which remains in place for common data)
- Caching is in a class now

### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/8](https://github.com/vindevoy/cherryblog/milestone/8)


# Release v1.2.0


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-17

## Enhancements

- Code is now following the Model-View-Controller pattern
- Logging updated for CherryPy and added for CherryBlog
- Updated the caching system, all data is stored in memory
- Server can now run in daemon mode, including user privileges
- Try Except added when reading configuration and content
- Texts are now in I8N directory, for future translations
- Static directories and static files can be served
- Separator is now a setting
- Minor bug fixes


### Github 

For more information on this release, see the issues for this milestone:

- [https://github.com/vindevoy/cherryblog/milestone/6](https://github.com/vindevoy/cherryblog/milestone/6)


# Release v1.1.1


- author: Yves Vindevogel (vindevoy)
- date: 2020-04-11

## Fixes

- Missing top image in the posts archive
- Alignment of the footer for small devices
- Code boxes were not showing nicely

### Github 


src/application/model/tags.py
###
#
#   Full history: see below
#
#   Version: 1.4.0
#   Date: 2020-05-08
#   Author: Yves Vindevogel (vindevoy)
#
#   Changes:
#       - Support for drafts
#       - Remapping of URLs to documents
#
###

import logging
import math
import string

from operator import itemgetter
from pathlib import Path

from common.content import Content
from common.datetime_support import DateTimeSupport
from common.options import Options
from common.singleton import Singleton
from common.tags_support import TagsSupport
from controller.remapper import Remapper


class Tags(metaclass=Singleton):
    __base_dir = 'tags'

    __logger = None

    def __init__(self):
        self.__logger = logging.getLogger('MODEL.TAGS')
        self.__logger.setLevel(Options().default_logging_level)

    def skip_tags(self):
        settings = Content().load_data_settings_yaml(self.__base_dir)
        self.__logger.debug('list - settings: {0}'.format(settings))

        tags = []

        for tag in settings['skip_tags']:
            tags.append(TagsSupport().tag_label(tag))

        return tags

    def list(self, posts):
        self.__logger.debug('list - posts: {0}'.format(posts))

        settings = Content().load_data_settings_yaml(self.__base_dir)
        self.__logger.debug('list - settings: {0}'.format(settings))

        # Starting with a dictionary as this is the easiest to find existing tags
        tags = {}

        for entry in posts:
            directory = entry['directory']
            file = entry['file']

            meta, _, _ = Content().read_content(directory, file)  # No need to catch the content

            if meta['tags'] is None:
                continue

            for tag in meta['tags']:
                label = TagsSupport().tag_label(tag)

                if label in settings['skip_tags']:
                    self.__logger.debug('list - tag {0} found in skip_tags'.format(tag))
                    continue

                if label in tags.keys():
                    self.__logger.debug('list - tag {0} already exists, +1'.format(tag))
                    current_count = tags[label]['count']
                    tags[label]['count'] = current_count + 1
                else:
                    self.__logger.debug('list - tag {0} does not already exist'.format(tag))
                    data = {'label': label, 'count': 1, 'text': string.capwords(tag)}
                    tags[label] = data

        self.__logger.debug('list - tags: '.format(tags))

        # Pushing this into a simple array for Jinja2
        tags_array = []

        for _, value in tags.items():  # Only need the value
            tags_array.append(value)

        for tag in tags_array:
            label = tag['label']
            unmapped = '/tags/{0}'.format(label)
            remapped = Remapper().remap_document(unmapped)

            if unmapped != remapped:
                label = remapped.split('/')[2]
                tag['label'] = label
                tag['text'] = TagsSupport().tag_text(label)

        tags_list = sorted(tags_array, key=itemgetter('count'), reverse=True)
        self.__logger.debug('list - sorted tags: '.format(tags_list))

        return tags_list

    def data(self, posts, tag, page_index, index_max_posts, count_tag_posts):
        self.__logger.debug('data - posts: {0}'.format(posts))
        self.__logger.debug('data - tag: {0}'.format(tag))
        self.__logger.debug('data - page_index tags: {0}'.format(page_index))
        self.__logger.debug('data - index_max_posts tags: {0}'.format(index_max_posts))
        self.__logger.debug('data - count_tag_posts tags: {0}'.format(count_tag_posts))

        data = {'posts': []}

        count_entries = 0
        max_entries = index_max_posts
        skip_entries = (int(page_index) - 1) * max_entries
        self.__logger.debug('data - max_entries: {0}'.format(max_entries))
        self.__logger.debug('data - skip_entries: {0}'.format(skip_entries))

        for entry in posts:
            directory = entry['directory']
            file = entry['file']

            post, _, post['content'] = Content().read_content(directory, file)

            if post['tags'] is None:
                continue

            self.__logger.debug('data - post: {0}'.format(post))

            must_include = False

            for tag_raw in post['tags']:
                if TagsSupport().tag_label(tag_raw) == tag:
                    must_include = True
                    break

            self.__logger.debug('data - must_include: {0}'.format(must_include))

            if must_include:
                count_entries += 1

                # We count the entries, but for pages 2 and more, you don't show them
                if skip_entries >= count_entries:
                    self.__logger.debug('data - post skipped}')
                    continue
                else:
                    self.__logger.debug('data - post added')

                stem = Path(file).stem
                url = '/posts/{0}'.format(stem)
                url = Remapper().remap_document(url)

                post['url'] = url
                post['date'] = DateTimeSupport().rewrite_date(post['date'])

                data['posts'].append(post)

                if count_entries == (max_entries + skip_entries):
                    self.__logger.debug('data - enough posts')
                    break

        data['tag'] = {'name': TagsSupport().tag_text(tag), 'path': tag}

        total_index_pages = math.ceil(count_tag_posts / max_entries)
        self.__logger.debug('data - total_index_pages: {0}'.format(total_index_pages))

        data['pagination'] = {'current_page': int(page_index), 'total_pages': total_index_pages}

        self.__logger.debug('data - {0}'.format(data))

        return data

    def count_posts(self, posts, tag):
        self.__logger.debug('count_posts - tag: {0}'.format(tag))
        self.__logger.debug('count_posts - posts: {0}'.format(posts))

        count_entries = 0

        for entry in posts:
            directory = entry['directory']
            file = entry['file']

            post, _, _ = Content().read_content(directory, file)

            if post['tags'] is None:
                continue

            for tag_raw in post['tags']:
                if TagsSupport().tag_label(tag_raw) == tag:
                    count_entries += 1
                    self.__logger.debug('count_posts - file {0} includes tag '.format(file))
                    break

        self.__logger.debug('count_posts - tag {0} has {1} posts'.format(tag, count_entries))

        return count_entries

###
#
#   Version: 1.3.0
#   Date: 2020-05-01
#   Author: Yves Vindevogel (vindevoy)
#
#   Changes:
#       - Moved the tag_label and tag_text methods to a support class in common
#       - Rewrite date format
#
#   Version: 1.2.1
#   Date: 2020-04-23
#   Author: Yves Vindevogel (vindevoy)
#
#   Fixes:
#       - Error on page with no tags
#
#   Version: 1.2.0
#   Date: 2020-04-17
#   Author: Yves Vindevogel (vindevoy)
#
#   Features:
#       - Caching done outside this class
#
#   Version: 1.1.0
#   Date: 2020-04-15
#   Author: Yves Vindevogel (vindevoy)
#
#   Features:
#       - Added logging
#
#   Version: 1.0.0
#   Date: 2020-04-13
#   Author: Yves Vindevogel (vindevoy)
#
#   This class was split of the DataLoader class
#
###
